src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,639 @@
+/*
+ * Copyright (c) 2023 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <openssl/opensslv.h>
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+
+#include "soter/soter_rsa_key.h"
+
+#include <string.h>
+
+#include <openssl/bn.h>
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/param_build.h>
+#include <openssl/rsa.h>
+
+#include "soter/openssl/soter_rsa_key_utils.h"
+#include "soter/soter_portable_endian.h"
+#include "soter/soter_wipe.h"
+
+soter_status_t soter_engine_specific_to_rsa_pub_key(const soter_engine_specific_rsa_key_t* engine_key,
+                                                    soter_container_hdr_t* key,
+                                                    size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    int rsa_mod_size;
+    size_t output_length;
+    uint32_t* pub_exp;
+    // This bigint is reused multiple times: read into it, serialize it, read into it...
+    BIGNUM* tmp = NULL;
+
+    if (!key_length) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (EVP_PKEY_RSA != EVP_PKEY_id(pkey)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    // Relevant pkey params
+    // OSSL_PKEY_PARAM_BITS, "bits", int
+    // OSSL_PKEY_PARAM_RSA_N, "n", big uint
+    // OSSL_PKEY_PARAM_RSA_E, "e", big uint
+
+    // See https://docs.cossacklabs.com/themis/spec/asymmetric-keypairs/rsa/ for key layout
+
+    if (!EVP_PKEY_get_int_param(pkey, OSSL_PKEY_PARAM_BITS, &rsa_mod_size)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+    rsa_mod_size = (rsa_mod_size + 7) / 8;
+
+    if (!is_mod_size_supported(rsa_mod_size)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    output_length = rsa_pub_key_size(rsa_mod_size);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_N, &tmp)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (BN_bn2binpad(tmp, (unsigned char*)(key + 1), rsa_mod_size) == -1) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_E, &tmp)) {
```

Comment:
I _think_ so. `man` doesn't say the opposite, but says
> If *bn is NULL then the BIGNUM is allocated by the method.

So it is possible to pass non-NULL pointer, and most logical behavior is to reuse passed one and put value there. Didn't go deep inside openssl code for this, but at least memory leak check passes on CI

Reply:
you conversation pushed to think that bignum is a value with dynamic length and about cases when a first call with `null` as argument will allocate buffer with size X and next call with non-null argument will require size > X... what openssl will do in that case? will it free and allocate new? will it fails? will it just allocate new without freeing? Will it often case when it fails due to not enough size of bignum? If so, can we allocate manually with the max required size of bignum enough for all parameters?
@iamnotacake , can you test case when with manual allocation of the bignum of small size and pass it as container for bigger value, for example RSA parameter? what will do openssl? and look on bignum sizes of the rsa parameters, which is the biggest?

Reply:
Looked inside of `EVP_PKEY_get_bn_param()` and turns out [it works a bit different than I thought](https://github.com/openssl/openssl/blob/master/crypto/evp/p_lib.c#L2141-L2181):
* Prepares on-stack buffer (2048 bytes, always enough for us)
* Prepares on-stack `OSSL_PARAM` to extract bigint (pointing to that on-stack buffer, will be in native-endian)
* Extracts value with `EVP_PKEY_get_params()`
* Creates/fills `BIGINT*` from temporary buffer
* Returns bigint to caller
* Seems like that temporary buffer is not erased :/

But since our goal is to serialize that bigint, here is what we could do instead, and simply skip allocating `BIGINT`:
* Prepare on-stack buffer (66 bytes max for EC P-521, 1024 bytes for RSA 8192)
* Read native-endian bigint into it using `EVP_PKEY_get_params()`
* Reverse bytes order while copying into serialized Themis key
* Cleanup: erase that buffer (if used for private key)
* Possible optimization: extract value directly into Themis key buffer on big-endian architectures

#### Answering question about how `EVP_PKEY_get_bn_param()` works when you pass it existing allocated BIGINT:
Deep under the hood internal function `bin2bn()` is used. It takes old pointer (be it `NULL` or some valid `BIGINT*`) and returns `BIGINT*`. In case prev ptr was `NULL`, it allocates one. If not, it just uses old `BIGINT*`. In case size is not large enough, it uses internal func `bn_wexpand()` to expand it. Which means it will modify `tmp` we passed, but not allocate completely separate one. In case expansion failed (out of memory?), old value is freed and NULL is returned, which means our `tmp` will also become `NULL`.

#### What if `tmp` is initialized, but not large enough?
Allocating `tmo` with `BN_new()` is OK. Although, this func doesn't reserve any space inside, so first time passing it to fill, inner buffer will be allocated, "resized" from zero to some big enough length.
Didn't find function to create `BIGINT*` with some specific capacity though.

Reply:
🤯

☝️ This is why I “love” OpenSSL.

Reply:
actually, I don't like the idea of bringing function of conversion between little and big endian, that looks strange in the openssl - https://github.com/openssl/openssl/blob/master/crypto/bn/bn_lib.c#L544
It should be tested somehow too

Reply:
> No, that function bn_wexpand is not documented, it's internal
Internal function is `bn_expand` and declared as static. And they place some of such things into the `internal` folder. So we can use this function to explicitly specify required size.
> And your ideas sounds a bit confusing to me.
What actually confuse you? We will do the same but explicitly allocate required size of bignum to be sure that it will not be resized internally and it's all.

> When serializing BIGINTs into Themis key, we only need BN_bn2bin*, we don't need little-endian variants there. We may need "little endian bytes to BIGINT" if we use EVP_PKEY_get_params() which fills raw buffer in native order.

Ok, so we will need only `bn2bin`. I don't remember which endianness used in openssl or themis and at which places.

Reply:
>  don't like the idea of bringing function of conversion between little and big endian

We don't need any conversion if we use `EVP_PKEY_get_bn_param()`, because
1) no matter how `EVP_PKEY_get_params()` behaves inside it, we get proper BIGINT
2) we later serialize BIGINT with `BN_bn2binpad()` so Themis key will contain proper big-endian byte order

Reply:
> Internal function is bn_expand and declared as static. And they place some of such things into the internal folder. So we can use this function to explicitly specify required size.

By internal I mean it's not public OpenSSL API and we should not use it

> I don't remember which endianness used in openssl or themis and at which places.

* Themis uses big
* `EVP_PKEY_get_params()` that fills provided buffer, uses native (usually little)
* `EVP_PKEY_get_bn_param()` returns BIGINT so we can later serialize it however we like
* `EVP_PKEY_fromdata()` uses native (usually little) if we pass bigint as buffer of bytes
* `EVP_PKEY_fromdata()` could also consume BIGINT if we use builder (OSSL_PARAM_BLD), _but that seems to just serialize our bigint in native order under the hood so it actually still consumes buffer but with proper byte order_